src/datetime/builder.rs

```diff
@@ -0,0 +1,159 @@
+use super::*;
+use std::convert::TryFrom;
+use time::Date;
+
+/// Builder for constructing a BSON [`DateTime`]
+pub struct DateTimeBuilder<Y = NoYear, M = NoMonth, D = NoDay> {
+    pub(crate) year: Y,
+    pub(crate) month: M,
+    pub(crate) day: D,
+
+    pub(crate) hour: Option<u8>,
+    pub(crate) minute: Option<u8>,
+    pub(crate) second: Option<u8>,
+    pub(crate) millisecond: Option<u16>,
+}
+
+impl Default for DateTimeBuilder {
+    fn default() -> Self {
+        Self {
+            year: NoYear,
+            month: NoMonth,
+            day: NoDay,
+            hour: None,
+            minute: None,
+            second: None,
+            millisecond: None,
+        }
+    }
+}
+
+pub struct Year(i32);
+pub struct NoYear;
+
+pub struct Month(u8);
+pub struct NoMonth;
+
+pub struct Day(u8);
+pub struct NoDay;
+
+impl<M, D> DateTimeBuilder<NoYear, M, D> {
+    /// Sets the year for the builder instance.
+    pub fn year(self, y: i32) -> DateTimeBuilder<Year, M, D> {
+        let Self {
+            year: _,
+            month,
+            day,
+            hour,
+            minute,
+            second,
+            millisecond,
+        } = self;
+        DateTimeBuilder {
+            year: Year(y),
+            month,
+            day,
+            hour,
+            minute,
+            second,
+            millisecond,
+        }
+    }
+}
+
+impl<Y, D> DateTimeBuilder<Y, NoMonth, D> {
+    /// Sets the month for the builder instance.
```

Comment:
These docstrings are good, but they are a little straightforward. I think it would be good to include some extra information that user might not necessarily be able to infer themselves, e.g. supported values (helps the user know if january is 1 or 0) or what happens when an invalid value is provided (the final `build` will return an error).

Ditto for others. For instance, in the `day` builder, it could be good to clarify what are accepted values to help answer questions like: what if we supply a day that's between 1 and 31 but isn't valid for that month year combo? (e.g. feb 29th on a non-leap year or the 31st of april.

Also, it's probably worth mentioning in the docstrings which ones are required and which ones are optional, and if one is optional what the default value will be.